### DISCUSSION TITLE
Cannot get this to work

### DISCUSSION THREAD

**Discussion by kerravon796** [xchwarze/wifi-pineapple-cloner#13] [General]
I have installed this a per instructions on ubuntu 20.04, had to install things like make and php when some commands errored.
the steps i followed are shown below.  I almost got it working but fails with missing packages like IW.  what other linux versions have people used? any one have a VM I can look at? thanks in advance.
kerravon

# Install necessary packages
 apt install git -y
 #apt install binwalk -y
 apt install php
 apt install make -y
 apt install python2.7 -y
 
# Install Binwalk
  git clone https://github.com/ReFirmLabs/binwalk.git
  cd binwalk
  sudo ./deps.sh
  sudo python ./setup.py install

 #Download Wifi Pineapple Cloner
 git clone https://github.com/xchwarze/wifi-pineapple-cloner.git
 cd wifi-pineapple-cloner

# Download Firmware and Extract it
# For tetra
 cd  build/fw-base
 wget https://www.wifipineapple.com/downloads/tetra/latest -O basefw.bin
 binwalk basefw.bin -e

# binwalk _basefw.bin.extracted/sysupgrade-pineapple-tetra/root -eM --run-as=root
 binwalk -eM --run-as=root  _basefw.bin.extracted/sysupgrade-pineapple-tetra/root
 mv _basefw.bin.extracted/sysupgrade-pineapple-tetra/_root.extracted/squashfs-root/ rootfs-base

# Get packages list
 php tools/opkg-parser.php rootfs-base/usr/lib/opkg/status

# Copy pineapple files
 chmod +x tools/copier.sh
 tools/copier.sh lists/tetra.filelist rootfs-base rootfs

# Fix files
 chmod +x tools/fs-patcher.sh
 tools/fs-patcher.sh mips tetra rootfs

# For this poc use openwrt imagebuilder v19.07.2 for ar71xx
 wget https://downloads.openwrt.org/releases/19.07.7/targets/ar71xx/generic/openwrt-imagebuilder-19.07.2-ar71xx-generic.Linux-x86_64.tar.xz
 tar xJf openwrt-imagebuilder-19.07.7-ar71xx-generic.Linux-x86_64.tar.xz
cd openwrt-imagebuilder-19.07.7-ar71xx-generic.Linux-x86_64

# based on step 2 data!
# ar71xx profile name: gl-ar150
# ath79 profile name: glinet_gl-ar150
make image PROFILE=gl-ar150 PACKAGES="at autossh base-files block-mount ca-certificates chat dnsmasq e2fsprogs ethtool firewall hostapd-utils ip6tables iperf3 iwinfo kmod-crypto-manager kmod-fs-ext4 kmod-fs-nfs kmod-fs-vfat kmod-gpio-button-hotplug kmod-ipt-offload kmod-leds-gpio kmod-ledtrig-default-on kmod-ledtrig-netdev kmod-ledtrig-timer kmod-mt76x2u kmod-nf-nathelper kmod-rt2800-usb kmod-rtl8187 kmod-rtl8192cu kmod-scsi-generic kmod-usb-acm kmod-usb-net-asix kmod-usb-net-asix-ax88179 kmod-usb-net-qmi-wwan kmod-usb-net-rndis kmod-usb-net-sierrawireless kmod-usb-net-smsc95xx kmod-usb-ohci kmod-usb-storage-extras kmod-usb-uhci kmod-usb2 libbz2-1.0 libcurl4 libelf1 libffi libgmp10 libiconv-full2 libintl libltdl7 libnet-1.2.x libnl200 libreadline8 libustream-mbedtls20150806 libxml2 logd macchanger mt7601u-firmware mtd nano ncat netcat nginx odhcp6c odhcpd-ipv6only openssh-client openssh-server openssh-sftp-server openssl-util php7-cgi php7-fpm php7-mod-hash php7-mod-json php7-mod-mbstring php7-mod-openssl php7-mod-session php7-mod-sockets php7-mod-sqlite3 ppp ppp-mod-pppoe procps-ng-pkill procps-ng-ps python-logging python-openssl python-sqlite3 rtl-sdr ssmtp tcpdump uboot-envtools uci uclibcxx uclient-fetch urandom-seed urngd usb-modeswitch usbreset usbutils wget wireless-tools wpad busybox libatomic1 libstdcpp6 -wpad-basic -dropbear" FILES=../rootfs
cp bin/targets/ar71xx/generic/openwrt-19.07.2-ar71xx-generic-gl-ar150-squashfs-sysupgrade.bin ../gl-ar150-pineapple-nano.bin

**Comment by kerravon796** [comment]
here is the errors produced by the above.                                                                                                                                                               

2024-02-23 04:58:39 (2.56 MB/s) - ‘basefw.bin’ saved [31989793]


Extractor Exception: Binwalk extraction uses many third party utilities, which may not be secure. If you wish to have extraction utilities executed as the current user, use '--run-as=root' (binwalk itself must be run as root).
----------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/binwalk-2.3.3+cddfede-py2.7.egg/binwalk/core/module.py", line 258, in __init__
    self.load()
  File "/usr/local/lib/python2.7/dist-packages/binwalk-2.3.3+cddfede-py2.7.egg/binwalk/modules/extractor.py", line 147, in load
    raise ModuleException("Binwalk extraction uses many third party utilities, which may not be secure. If you wish to have extraction utilities executed as the current user, use '--run-as=%s' (binwalk itself must be run as root)." % user_info.pw_name)
ModuleException: Binwalk extraction uses many third party utilities, which may not be secure. If you wish to have extraction utilities executed as the current user, use '--run-as=root' (binwalk itself must be run as root).
----------------------------------------------------------------------------------------------------


General Error: Cannot open file _basefw.bin.extracted/sysupgrade-pineapple-tetra/root (CWD: /home/iot/binwalk/wifi-pineapple-cloner/build/fw-base) : [Errno 2] No such file or directory: '_basefw.bin.extracted/sysupgrade-pineapple-tetra/root'

mv: cannot stat '_basefw.bin.extracted/sysupgrade-pineapple-tetra/_root.extracted/squashfs-root/': No such file or directory
./autobuild4.sh: line 32: php: command not found
chmod: cannot access 'tools/copier.sh': No such file or directory
./autobuild4.sh: line 36: tools/copier.sh: No such file or directory
chmod: cannot access 'tools/fs-patcher.sh': No such file or directory
./autobuild4.sh: line 40: tools/fs-patcher.sh: No such file or directory
--2024-02-23 04:58:39--  https://downloads.openwrt.org/releases/19.07.7/targets/ar71xx/generic/openwrt-imagebuilder-19.07.2-ar71xx-generic.Linux-x86_64.tar.xz
Resolving downloads.openwrt.org (downloads.openwrt.org)... 199.232.54.132, 2a04:4e42:4a::644
Connecting to downloads.openwrt.org (downloads.openwrt.org)|199.232.54.132|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-02-23 04:58:40 ERROR 404: Not Found.

tar (child): openwrt-imagebuilder-19.07.7-ar71xx-generic.Linux-x86_64.tar.xz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
./autobuild4.sh: line 45: cd: openwrt-imagebuilder-19.07.7-ar71xx-generic.Linux-x86_64: No such file or directory
Enter the profile name:
gl-ar750s
make: *** No rule to make target 'image'.  Stop.
cp: cannot stat 'bin/targets/ar71xx/generic/openwrt-19.07.2-ar71xx-generic-gl-ar750s-squashfs-sysupgrade.bin': No such file or directory
root@attifyos:/home/iot#

**Comment by xchwarze** [comment]
From builder vm:
```bash
dsr@dsr-virtual-machine:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.4 LTS
Release:	20.04
Codename:	focal

dsr@dsr-virtual-machine:~$ python3 --version
Python 3.8.10

dsr@dsr-virtual-machine:~$ binwalk -h

Binwalk v2.3.3
Craig Heffner, ReFirmLabs
https://github.com/ReFirmLabs/binwalk

Usage: binwalk [OPTIONS] [FILE1] [FILE2] [FILE3] ...
```

I recommend you to uninstall binwalk and install everything with the following steps
https://github.com/xchwarze/wifi-pineapple-cloner/blob/master/build.md#automatic-build-steps

I also noticed that you are getting a 404 download from the builder because you are getting the url wrong. For this I also recommend that you do it from the automatic tool:
```bash
dsr@dsr-virtual-machine:~/wifi-pineapple-cloner$ tools/dependencies-install.sh openwrt-deps-mips
Wifi Pineapple Cloner - dependencies
************************************** by DSR!

Install OpenWrt MIPS deps...
******************************
[*] Install: ar71xx-generic
    [+] Downloading imagebuilder...
        https://downloads.openwrt.org/releases/19.07.7/targets/ar71xx/generic/openwrt-imagebuilder-19.07.7-ar71xx-generic.Linux-x86_64.tar.xz
    [+] Install imagebuilder...
    [+] Correct opkg feeds
[*] Install: ar71xx-nand
    [+] Downloading imagebuilder...
        https://downloads.openwrt.org/releases/19.07.7/targets/ar71xx/nand/openwrt-imagebuilder-19.07.7-ar71xx-nand.Linux-x86_64.tar.xz
    [+] Install imagebuilder...
    [+] Correct opkg feeds
[*] Install: ath79-generic
    [+] Downloading imagebuilder...
        https://downloads.openwrt.org/releases/19.07.7/targets/ath79/generic/openwrt-imagebuilder-19.07.7-ath79-generic.Linux-x86_64.tar.xz
    [+] Install imagebuilder...
    [+] Correct opkg feeds
[*] Install: lantiq-xrx200
    [+] Downloading imagebuilder...
        https://downloads.openwrt.org/releases/19.07.7/targets/lantiq/xrx200/openwrt-imagebuilder-19.07.7-lantiq-xrx200.Linux-x86_64.tar.xz
    [+] Install imagebuilder...
    [+] Correct opkg feeds

[*] Install script end!
```

as can be seen in the log they are installed correctly.

Since I had cloned the repo and installed the imagebuilder I also ran the build command to check if everything was ok:
```bash
dsr@dsr-virtual-machine:~/wifi-pineapple-cloner$ tools/builder.sh mips universal imagebuilder/19.07.7-ar71xx-generic gl-ar750s
Wifi Pineapple Cloner - builder
************************************** by DSR!

[*] Checking dependencies
******************************
[!] Attention!

File 'build/fw-base/basefw.bin' was not found.
You must download the original TETRA firmware version 2.7.0 and place it with the name basefw.bin in this path:
/home/dsr/wifi-pineapple-cloner/build/fw-base

dsr@dsr-virtual-machine:~/wifi-pineapple-cloner$ cd build/fw-base/
dsr@dsr-virtual-machine:~/wifi-pineapple-cloner/build/fw-base$ wget https://www.wifipineapple.com/downloads/tetra/latest -O basefw.bin
--2024-02-23 13:51:14--  https://www.wifipineapple.com/downloads/tetra/latest
Resolving www.wifipineapple.com (www.wifipineapple.com)... 172.67.137.197, 104.21.46.105, 2606:4700:3031::ac43:89c5, ...
Connecting to www.wifipineapple.com (www.wifipineapple.com)|172.67.137.197|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/octet-stream]
Saving to: ‘basefw.bin’

basefw.bin              [        <=>         ]  30,51M  16,8MB/s    in 1,8s    

2024-02-23 13:51:17 (16,8 MB/s) - ‘basefw.bin’ saved [31989793]

dsr@dsr-virtual-machine:~/wifi-pineapple-cloner/build/fw-base$ cd ..
dsr@dsr-virtual-machine:~/wifi-pineapple-cloner/build$ cd ..
dsr@dsr-virtual-machine:~/wifi-pineapple-cloner$ tools/builder.sh mips universal imagebuilder/19.07.7-ar71xx-generic gl-ar750s
Wifi Pineapple Cloner - builder
************************************** by DSR!

[*] Checking dependencies
******************************
  [+] Check completed

[*] Prepare builder
******************************
  [+] Install: libubus20191227_2019-12-27-041c9d1c-1
  [+] Install: busybox_1.30.1-6
  [+] Builder setup complete

[*] Prepare build
******************************
  [+] Unpack firmware for get file system

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             POSIX tar archive (GNU), owner user name: "de-pineapple-tetra/"


DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Squashfs filesystem, little endian, version 4.0, compression:xz, size: 13757048 bytes, 2553 inodes, blocksize: 262144 bytes, created: 2020-02-27 21:05:12

  [+] Copying the original files
 Filelist2Copy - by DSR!
******************************

  [*] Start copy loop
  [*] Clean 'new fs' folder
  [*] Create 'new fs' folder
  [+] Files copied: 65

  [+] Patch file system
Wifi Pineapple Cloner v4
by DSR!
******************************

[*] Device detection fix
[*] Correct OPKG feed url
[*] Install panel fixes and improvements
[*] Enable ssh by default
[*] Change root password to: root
[*] Fix uci-defaults
[*] Fix pendrive hotplug
[*] Add support for reflash
[*] Fix airmon-ng listInterfaces()
[*] Add wpc-tools and service
[*] Other fixs
[*] Patch elf files
[*] Setting target as: universal
[*] Done!


[*] Build
******************************
  [+] Clean last build data
  [+] Executing make
Package list missing or not up-to-date, generating it.

Building package index...
Downloading http://downloads.openwrt.org/releases/19.07.7/targets/ar71xx/generic/packages/Packages.gz
Updated list of available packages in /home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/build_dir/target-mips_24kc_musl/root-ar71xx/../../../../../../../../home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/dl/openwrt_core
Downloading http://downloads.openwrt.org/releases/19.07.7/packages/mips_24kc/base/Packages.gz
Updated list of available packages in /home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/build_dir/target-mips_24kc_musl/root-ar71xx/../../../../../../../../home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/dl/openwrt_base
Downloading http://downloads.openwrt.org/releases/19.07.7/packages/mips_24kc/packages/Packages.gz
Updated list of available packages in /home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/build_dir/target-mips_24kc_musl/root-ar71xx/../../../../../../../../home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/dl/openwrt_packages
Downloading http://downloads.openwrt.org/releases/19.07.7/packages/mips_24kc/routing/Packages.gz
Updated list of available packages in /home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/build_dir/target-mips_24kc_musl/root-ar71xx/../../../../../../../../home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/dl/openwrt_routing
Downloading file:packages/Packages
Updated list of available packages in /home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/build_dir/target-mips_24kc_musl/root-ar71xx/../../../../../../../../home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/dl/imagebuilder
24+1 records in
25+0 records out
1638400 bytes (1.6 MB, 1.6 MiB) copied, 0.00114012 s, 1.4 GB/s
Pseudo file "/dev" exists in source filesystem "/home/dsr/wifi-pineapple-cloner/imagebuilder/19.07.7-ar71xx-generic/build_dir/target-mips_24kc_musl/root-ar71xx/dev".
Ignoring, exclude it (-e/-ef) to override.
3104+1 records in
3104+1 records out
1589481 bytes (1.6 MB, 1.5 MiB) copied, 0.00409962 s, 388 MB/s
24+1 records in
25+0 records out
1638400 bytes (1.6 MB, 1.6 MiB) copied, 0.000638499 s, 2.6 GB/s
20981+1 records in
20981+1 records out
10742432 bytes (11 MB, 10 MiB) copied, 0.0235442 s, 456 MB/s
padding image to 00bd0000
padding image to 00a40000
padding image to 00a3f000
padding image to 00a40000
82+1 records in
83+0 records out
10878976 bytes (11 MB, 10 MiB) copied, 0.0043457 s, 2.5 GB/s

[*] Firmware folder: /home/dsr/wifi-pineapple-cloner/build/release
******************************
total 37392
-rw-rw-r-- 1 dsr dsr    39249 feb 23 13:54 make.log
-rw-rw-r-- 1 dsr dsr     5273 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-device-gl-ar750s.manifest
-rw-rw-r-- 1 dsr dsr 12386537 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-gl-ar750s-squashfs-sysupgrade.bin
-rw-rw-r-- 1 dsr dsr 10878976 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-root.squashfs
-rw-rw-r-- 1 dsr dsr  1589725 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-uImage-lzma.bin
-rwxr-xr-x 1 dsr dsr  5059140 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-vmlinux.bin
-rwxr-xr-x 1 dsr dsr  5064264 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-vmlinux.elf
-rw-rw-r-- 1 dsr dsr  1638400 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-vmlinux.lzma
-rwxr-xr-x 1 dsr dsr  1599864 feb 23 13:54 openwrt-19.07.7-ar71xx-generic-vmlinux-lzma.elf
-rw-rw-r-- 1 dsr dsr     1035 feb 23 13:54 profiles.json
-rw-rw-r-- 1 dsr dsr     1075 feb 23 13:54 sha256sums
```

as you can see I tried 2 out of 3 commands and everything works correctly.

**Reply by kerravon796** [reply]
ok I will start with a new VM and follow above. will let you know How I get on.

**Comment by kerravon796** [comment]
Ok Installed it on ubuntu 23.04, had to install some missing tools when requested, python 2.7 didnt have any install path, downloaded it and installed that way. wifi-pineapple-cloner still errored towards the end. installed python is python3 and bingo everything works fine now.
thanks for all the info above helped me a lot and for your help in this.  BTW have you tried the latest model of the wifi pineapple?

**Comment by xchwarze** [comment]
They already deleted this repository once, so next time I guess they won't be satisfied with that. So I don't plan to do anything with that platform.

**Comment by kerravon796** [comment]
Ok, just found a document about the mark7 using either a Dlink DIR-509
or a D-Link_DIR-515 .  If I can get hold of either I might have a play with it.
Any many thanks for the help.
